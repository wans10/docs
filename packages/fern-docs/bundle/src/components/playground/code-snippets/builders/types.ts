import type { EndpointContext } from "@fern-api/fdr-sdk/api-definition";
import {
  buildEndpointUrl,
  wrapOpenRPCRequest,
} from "@fern-api/fdr-sdk/api-definition";

import {
  PlaygroundAuthState,
  PlaygroundEndpointRequestFormState,
} from "../../types";

export abstract class PlaygroundCodeSnippetBuilder {
  protected url: string;

  constructor(
    protected context: EndpointContext,
    protected formState: PlaygroundEndpointRequestFormState,
    protected authState: PlaygroundAuthState,
    protected baseUrl: string | undefined,
    protected redacted: boolean
  ) {
    // TODO: wire through the environment from hook
    this.url = buildEndpointUrl({
      endpoint: context.endpoint,
      pathParameters: formState.pathParameters,
      baseUrl,
    });
  }

  protected maybeWrapJsonBody(body: unknown): unknown {
    if (this.context.endpoint.protocol?.type === "openrpc") {
      return wrapOpenRPCRequest(
        body,
        this.context.endpoint.protocol.methodName
      );
    }
    return body;
  }

  public abstract build(): string;
}
