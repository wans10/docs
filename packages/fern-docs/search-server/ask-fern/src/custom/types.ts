export type CustomAskFernConfig = {
  isCohere: boolean;
};
