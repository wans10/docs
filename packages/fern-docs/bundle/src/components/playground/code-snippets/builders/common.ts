import { type EndpointDefinition } from "@fern-api/fdr-sdk/api-definition";
import { unknownToString } from "@fern-api/ui-core-utils";

export function buildPath(
  path: EndpointDefinition["path"],
  pathParameters?: Record<string, unknown>
): string {
  return path
    .map((part) => {
      if (part.type === "pathParameter") {
        const stateValue = unknownToString(pathParameters?.[part.value]);
        return stateValue.length > 0
          ? encodeURIComponent(stateValue)
          : ":" + part.value;
      }
      return part.value;
    })
    .join("");
}

export function buildUrlWithQueryParams(
  baseUrl: string,
  queryParameters: Record<string, unknown>
): string {
  const queryParams = Object.entries(queryParameters)
    .filter(([_, value]) => value != null)
    .map(([key, value]) => {
      if (Array.isArray(value)) {
        return value
          .map(
            (v) => `${encodeURIComponent(key)}=${encodeURIComponent(String(v))}`
          )
          .join("&");
      }
      return `${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`;
    })
    .join("&");

  return queryParams ? `${baseUrl}?${queryParams}` : baseUrl;
}

export function indentAfter(
  str: string,
  indent: number,
  afterLine?: number
): string {
  return str
    .split("\n")
    .map((line, idx) => {
      if (afterLine == null || idx > afterLine) {
        return " ".repeat(indent) + line;
      }
      return line;
    })
    .join("\n");
}
