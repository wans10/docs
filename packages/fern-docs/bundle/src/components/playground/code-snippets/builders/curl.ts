import {
  SnippetHttpRequest,
  SnippetHttpRequestBodyFormValue,
  convertToCurl,
} from "@fern-api/fdr-sdk/api-definition";
import { visitDiscriminatedUnion } from "@fern-api/ui-core-utils";

import { convertPlaygroundFormDataEntryValueToResolvedExampleEndpointRequest } from "../../types";
import { PlaygroundCodeSnippetBuilder } from "./types";

export class CurlSnippetBuilder extends PlaygroundCodeSnippetBuilder {
  private isFileForgeHackEnabled = false;

  public setFileForgeHackEnabled(isFileForgeHackEnabled: boolean): this {
    this.isFileForgeHackEnabled = isFileForgeHackEnabled;
    return this;
  }

  public override build(): string {
    // If the endpoint protocol is OpenRPC, ensure we add the application/json Content-Type header
    if (this.context.endpoint.protocol?.type === "openrpc") {
      // Create a copy of the headers to avoid mutating the original object
      const headers = { ...this.formState.headers };
      // Add or override the Content-Type header
      headers["Content-Type"] = "application/json";
      this.formState = {
        ...this.formState,
        headers,
      };
    }

    return convertToCurl(
      {
        method: this.context.endpoint.method,
        url: this.url,
        searchParams: this.formState.queryParameters,
        headers: this.formState.headers,
        basicAuth:
          this.context.auth?.type === "basicAuth"
            ? this.authState.basicAuth
            : undefined,
        body: this.#convertFormStateToBody(),
        redacted: this.redacted,
      },
      { usesApplicationJsonInFormDataValue: this.isFileForgeHackEnabled }
    );
  }

  #convertFormStateToBody(): SnippetHttpRequest["body"] {
    if (this.formState.body == null) {
      return undefined;
    }
    return visitDiscriminatedUnion(this.formState.body, "type")._visit<
      SnippetHttpRequest["body"]
    >({
      json: ({ value }) => ({
        type: "json",
        value: this.maybeWrapJsonBody(value),
      }),
      "form-data": ({ value }) => {
        const newValue: Record<string, SnippetHttpRequestBodyFormValue> = {};
        for (const [key, v] of Object.entries(value)) {
          const convertedV =
            convertPlaygroundFormDataEntryValueToResolvedExampleEndpointRequest(
              v
            );
          if (convertedV != null) {
            newValue[key] = convertedV;
          }
        }
        if (Object.keys(newValue).length === 0) {
          return undefined;
        }
        return { type: "form", value: newValue };
      },
      "octet-stream": ({ value }) =>
        value != null ? { type: "bytes", filename: value.name } : undefined,
      _other: () => undefined,
    });
  }
}
