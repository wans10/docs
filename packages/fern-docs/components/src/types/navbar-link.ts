export interface DefaultNavbarLink {
  type: "filled" | "outlined" | "minimal" | "primary" | "secondary";
  href: string;
  text: string | undefined;
  icon: string | undefined;
  rightIcon: string | undefined;
  rounded: boolean | undefined;
  className: string | undefined;
  id: string | undefined;
  returnToQueryParam: string | undefined;
}

export interface GithubNavbarLink {
  type: "github";
  href: string;
  className: string | undefined;
  id: string | undefined;
}

export type NavbarLink = DefaultNavbarLink | GithubNavbarLink;
