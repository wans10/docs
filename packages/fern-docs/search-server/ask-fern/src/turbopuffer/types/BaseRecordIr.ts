import { RecordIr } from "./RecordIr";

export type BaseRecordIr = Omit<RecordIr, "attributes"> & {
  attributes: Omit<RecordIr["attributes"], "chunk">;
};
