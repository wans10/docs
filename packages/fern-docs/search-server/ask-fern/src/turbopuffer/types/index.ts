export * from "./BaseRecordIr";
export * from "./TurbopufferRecord";
export * from "./RecordIr";
export * from "./ResultItem";
export * from "./TbufAttributeData";
