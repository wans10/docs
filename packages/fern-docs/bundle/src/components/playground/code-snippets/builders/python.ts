import { isNonNullish, visitDiscriminatedUnion } from "@fern-api/ui-core-utils";

import { PlaygroundFormDataEntryValue } from "../../types";
import { buildPath, buildUrlWithQueryParams, indentAfter } from "./common";
import { PlaygroundCodeSnippetBuilder } from "./types";

interface PythonRequestParams {
  json?: string;
  data?: string;
  files?: string;
}

export class PythonRequestSnippetBuilder extends PlaygroundCodeSnippetBuilder {
  #buildRequests({ json, data, files }: PythonRequestParams) {
    return `# ${this.context.node.title} (${this.context.endpoint.method} ${buildPath(this.context.endpoint.path)})
response = requests.${this.context.endpoint.method.toLowerCase()}(
  "${buildUrlWithQueryParams(this.url, this.formState.queryParameters)}",
  headers=${indentAfter(JSON.stringify({ ...this.formState.headers, "Content-Type": undefined }, undefined, 2), 2, 0)},${json != null ? `\n  json=${indentAfter(json, 2, 0)},` : ""}${
    data != null ? `\n  data=${indentAfter(data, 2, 0)},` : ""
  }${files != null ? `\n  files=${indentAfter(files, 2, 0)},` : ""}
)

print(response.json())`;
  }

  #formatFormDataValue(value: unknown): string {
    if (typeof value === "object" && !Array.isArray(value)) {
      return `json.dumps(${indentAfter(JSON.stringify(value, undefined, 2), 2, 0)})`;
    }
    if (
      typeof value === "object" &&
      Array.isArray(value) &&
      value.length === 1
    ) {
      return JSON.stringify(value[0]);
    }
    return JSON.stringify(value);
  }

  public override build(): string {
    const imports = ["requests"];

    if (this.formState.body == null) {
      return `${imports.map((pkg) => `import ${pkg}`).join("\n")}

${this.#buildRequests({})}`;
    }

    return visitDiscriminatedUnion(this.formState.body, "type")._visit<string>({
      json: ({ value }) => `${imports.map((pkg) => `import ${pkg}`).join("\n")}

${this.#buildRequests({ json: JSON.stringify(this.maybeWrapJsonBody(value), undefined, 2) })}`,
      "form-data": ({ value }) => {
        const singleFiles = Object.entries(value)
          .filter(
            (
              entry
            ): entry is [string, PlaygroundFormDataEntryValue.SingleFile] =>
              PlaygroundFormDataEntryValue.isSingleFile(entry[1])
          )
          .map(([k, v]) => {
            if (v.value == null) {
              return undefined;
            }
            return `'${k}', ('${v.value.name}', open('${v.value.name}', 'rb')),`;
          })
          .filter(isNonNullish);
        const fileArrays = Object.entries(value)
          .filter(
            (
              entry
            ): entry is [string, PlaygroundFormDataEntryValue.MultipleFiles] =>
              PlaygroundFormDataEntryValue.isMultipleFiles(entry[1])
          )
          .map(([k, v]) => {
            const fileStrings = v.value.map(
              (file) => `'${k}': ('${file.name}', open('${file.name}', 'rb'))`
            );
            if (fileStrings.length === 0) {
              return;
            }
            return fileStrings.join(",\n");
          })
          .filter(isNonNullish);

        const dataEntries = Object.entries(value)
          .filter(
            (entry): entry is [string, PlaygroundFormDataEntryValue.Json] =>
              PlaygroundFormDataEntryValue.isJson(entry[1]) ||
              PlaygroundFormDataEntryValue.isExploded(entry[1])
          )
          .map(([k, v]) => {
            if (PlaygroundFormDataEntryValue.isExploded(v)) {
              const items: string[] = [];
              for (const item of (v as PlaygroundFormDataEntryValue.Exploded)
                .value) {
                items.push(`'${k}': ${this.#formatFormDataValue(item)},`);
              }
              return items.join("\n");
            }

            return v.value == null
              ? undefined
              : `'${k}': ${this.#formatFormDataValue(v.value)},`;
          })
          .filter(isNonNullish)
          .join("\n");

        let fileEntries = [...singleFiles, ...fileArrays].join("\n");
        let data: string | undefined = undefined;

        if (fileEntries.length === 0 && dataEntries.length > 0) {
          fileEntries = Object.entries(value)
            .filter(
              (entry): entry is [string, PlaygroundFormDataEntryValue.Json] =>
                PlaygroundFormDataEntryValue.isJson(entry[1]) ||
                PlaygroundFormDataEntryValue.isExploded(entry[1])
            )
            .map(([k, v]) => {
              if (PlaygroundFormDataEntryValue.isExploded(v)) {
                const items: string[] = [];
                for (const item of (v as PlaygroundFormDataEntryValue.Exploded)
                  .value) {
                  items.push(
                    `'${k}': (None, ${this.#formatFormDataValue(item)})`
                  );
                }
                return items.join(",\n");
              }

              return v.value == null
                ? undefined
                : `'${k}': (None, ${this.#formatFormDataValue(v.value)})`;
            })
            .filter(isNonNullish)
            .join(",\n");
        } else if (dataEntries.length > 0) {
          data = `{\n${indentAfter(dataEntries, 2)}\n}`;
        }

        const files =
          fileEntries.length > 0
            ? `{\n${indentAfter(fileEntries, 2)}\n}`
            : undefined;

        if (
          data?.includes("json.dumps") ||
          fileEntries.includes("json.dumps")
        ) {
          imports.push("json");
        }

        return `${imports.map((pkg) => `import ${pkg}`).join("\n")}

${this.#buildRequests({ data, files })}`;
      },
      "octet-stream": (
        f
      ) => `${imports.map((pkg) => `import ${pkg}`).join("\n")}

${this.#buildRequests({ data: f.value != null ? `open('${f.value?.name}', 'rb').read()` : undefined })}`,
      _other: () => `${imports.map((pkg) => `import ${pkg}`).join("\n")}

${this.#buildRequests({})}`,
    });
  }
}
