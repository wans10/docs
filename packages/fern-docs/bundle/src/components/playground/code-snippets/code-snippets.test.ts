import {
  EndpointContext,
  EndpointDefinition,
  EndpointId,
  EnvironmentId,
  PropertyKey,
} from "@fern-api/fdr-sdk/api-definition";
import {
  ApiDefinitionId,
  EndpointNode,
  NodeId,
  Slug,
} from "@fern-api/fdr-sdk/navigation";

import { PlaygroundEndpointRequestFormState } from "../types";
import { CurlSnippetBuilder } from "./builders/curl";
import { PythonRequestSnippetBuilder } from "./builders/python";
import { TypescriptFetchSnippetBuilder } from "./builders/typescript";

// Mock File class for testing
class MockFile {
  constructor(
    public readonly name: string,
    public readonly type: string,
    public readonly lastModified: number = Date.now()
  ) {}

  // Required File interface properties
  public readonly size = 0;
  public readonly webkitRelativePath = "";
  public readonly arrayBuffer = async () => new ArrayBuffer(0);
  public readonly slice = () => new Blob();
  public readonly stream = () => new ReadableStream();
  public readonly text = async () => "";
  public readonly bytes = async () => new Uint8Array(0);
}

// mock file option
const mockFile = (name: string, type: string): MockFile => {
  return new MockFile(name, type);
};

describe("PlaygroundCodeSnippetBuilder", () => {
  const node: EndpointNode = {
    type: "endpoint",
    method: "POST",
    endpointId: EndpointId(""),
    isResponseStream: undefined,
    playground: undefined,
    title: "My endpoint",
    slug: Slug(""),
    canonicalSlug: undefined,
    icon: undefined,
    hidden: undefined,
    id: NodeId(""),
    apiDefinitionId: ApiDefinitionId(""),
    availability: undefined,
    authed: undefined,
    viewers: undefined,
    orphaned: undefined,
    featureFlags: undefined,
  };

  const endpoint: EndpointDefinition = {
    id: EndpointId(""),
    auth: undefined,
    protocol: undefined,
    availability: undefined,
    defaultEnvironment: EnvironmentId("Prod"),
    environments: [
      {
        id: EnvironmentId("Prod"),
        baseUrl: "https://example.com",
      },
    ],
    displayName: "endpoint-1",
    operationId: "endpoint-1",
    method: "POST",
    path: [
      { type: "literal", value: "/test/" },
      { value: PropertyKey("test"), type: "pathParameter" },
    ],
    pathParameters: [
      {
        key: PropertyKey("test"),
        valueShape: {
          type: "alias",
          value: {
            type: "primitive",
            value: {
              type: "string",
              format: undefined,
              regex: undefined,
              minLength: undefined,
              maxLength: undefined,
              default: undefined,
            },
          },
        },
        description: undefined,
        availability: undefined,
      },
    ],
    queryParameters: undefined,
    requestHeaders: undefined,
    requests: undefined,
    responses: undefined,
    errors: [],
    examples: [],
    snippetTemplates: undefined,
    description: undefined,
    responseHeaders: undefined,
    namespace: undefined,
  };
  const formState: PlaygroundEndpointRequestFormState = {
    type: "endpoint",
    headers: {
      Accept: "application/json",
      Test: "test",
    },
    pathParameters: {
      test: "hello@example",
    },
    queryParameters: {},
    body: {
      type: "json",
      value: {
        test: "hello",
        deeply: {
          nested: 1,
        },
      },
    },
  };

  const context: EndpointContext = {
    node,
    endpoint,
    auth: undefined,
    types: {},
    globalHeaders: [],
  };

  const multipartFormState: PlaygroundEndpointRequestFormState = {
    type: "endpoint",
    headers: {
      Accept: "application/json",
      "Content-Type": "multipart/form-data",
    },
    pathParameters: {
      test: "hello@example",
    },
    queryParameters: {},
    body: {
      type: "form-data",
      value: {
        file1: {
          type: "file",
          value: mockFile("test.txt", "text/plain"),
        },
        file2: {
          type: "fileArray",
          value: [
            mockFile("image1.jpg", "image/jpeg"),
            mockFile("image2.jpg", "image/jpeg"),
          ],
        },
        textField: {
          type: "json",
          value: "simple string value",
        },
        jsonField: {
          type: "json",
          value: {
            nested: {
              value: 123,
            },
          },
        },
      },
    },
  };

  const multipartFormStateWithoutFiles: PlaygroundEndpointRequestFormState = {
    type: "endpoint",
    headers: {
      Accept: "application/json",
      "Content-Type": "multipart/form-data",
    },
    pathParameters: {
      test: "hello@example",
    },
    queryParameters: {},
    body: {
      type: "form-data",
      value: {
        textField: {
          type: "json",
          value: "simple string value",
        },
        jsonField: {
          type: "json",
          value: {
            nested: {
              value: 123,
            },
          },
        },
      },
    },
  };

  it("should render curl", () => {
    expect(
      new CurlSnippetBuilder(context, formState, {}, undefined, false).build()
    ).toMatchSnapshot();
  });

  it("should render python", () => {
    expect(
      new PythonRequestSnippetBuilder(
        context,
        formState,
        {},
        undefined,
        false
      ).build()
    ).toMatchSnapshot();
  });

  it("should render typescript", () => {
    expect(
      new TypescriptFetchSnippetBuilder(
        context,
        formState,
        {},
        undefined,
        false
      ).build()
    ).toMatchSnapshot();
  });

  it("should render python with multipart form data", () => {
    expect(
      new PythonRequestSnippetBuilder(
        context,
        multipartFormState,
        {},
        undefined,
        false
      ).build()
    ).toMatchSnapshot();
  });

  it("should render python with multipart form data without files", () => {
    expect(
      new PythonRequestSnippetBuilder(
        context,
        multipartFormStateWithoutFiles,
        {},
        undefined,
        false
      ).build()
    ).toMatchSnapshot();
  });
});
