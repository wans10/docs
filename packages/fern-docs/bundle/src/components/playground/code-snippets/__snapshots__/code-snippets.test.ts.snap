// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlaygroundCodeSnippetBuilder > should render curl 1`] = `
"curl -X POST https://example.com/test/hello%40example \\
     -H "Accept: application/json" \\
     -H "Test: test" \\
     -d '{
  "test": "hello",
  "deeply": {
    "nested": 1
  }
}'"
`;

exports[`PlaygroundCodeSnippetBuilder > should render python 1`] = `
"import requests

# My endpoint (POST /test/:test)
response = requests.post(
  "https://example.com/test/hello%40example",
  headers={
    "Accept": "application/json",
    "Test": "test"
  },
  json={
    "test": "hello",
    "deeply": {
      "nested": 1
    }
  },
)

print(response.json())"
`;

exports[`PlaygroundCodeSnippetBuilder > should render python with multipart form data 1`] = `
"import requests
import json

# My endpoint (POST /test/:test)
response = requests.post(
  "https://example.com/test/hello%40example",
  headers={
    "Accept": "application/json"
  },
  data={
    'textField': "simple string value",
    'jsonField': json.dumps({
        "nested": {
          "value": 123
        }
      }),
  },
  files={
    'file1', ('test.txt', open('test.txt', 'rb')),
    'file2': ('image1.jpg', open('image1.jpg', 'rb')),
    'file2': ('image2.jpg', open('image2.jpg', 'rb'))
  },
)

print(response.json())"
`;

exports[`PlaygroundCodeSnippetBuilder > should render python with multipart form data without files 1`] = `
"import requests
import json

# My endpoint (POST /test/:test)
response = requests.post(
  "https://example.com/test/hello%40example",
  headers={
    "Accept": "application/json"
  },
  files={
    'textField': (None, "simple string value"),
    'jsonField': (None, json.dumps({
        "nested": {
          "value": 123
        }
      }))
  },
)

print(response.json())"
`;

exports[`PlaygroundCodeSnippetBuilder > should render typescript 1`] = `
"// My endpoint (POST /test/:test)
const response = await fetch("https://example.com/test/hello%40example", {
  method: "POST",
  headers: {
    "Accept": "application/json",
    "Test": "test"
  },
  body: JSON.stringify({
    "test": "hello",
    "deeply": {
      "nested": 1
    }
  }),
});

const body = await response.json();
console.log(body);"
`;
